Remove finput and fgrad_input from slow3d transpose signatures

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69584
* #69044
* __->__ #68899

Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

jbschlosser committed:
Remove finput and fgrad_input from slow3d transpose signatures

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).

albanD reviewed (commented):
The code update sounds good.
Are the previous benchmark still relevant? What's the impact of this PR?

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).

jbschlosser commented on aten/src/ATen/native/NaiveConvolutionTranspose3d.cpp:
```diff
@@ -278,21 +271,12 @@ void slow_conv_transpose3d_out_cpu_template(
-                   input_depth * input_height * input_width});
-  columns.zero_();
+  // Create temporary columns
+  Tensor columns = at::empty({n_output_plane * kernel_width * kernel_height * kernel_depth,
```
I think empty is better here because it's overwritten when the columns are extracted below

albanD replied:
So the original code was setting everything to 0 wrongfully?

jbschlosser replied:
Yep that's correct, no need to set to 0. `columns` is first used below:

```cpp
cpublas::gemm(
  TransposeType::NoTranspose,
  TransposeType::Transpose,
  n,
  m,
  k,
  1,
  input_n.data_ptr<scalar_t>(),
  n,
  weight.data_ptr<scalar_t>(),
  m,
  0,
  columns.data_ptr<scalar_t>(),
  n);
```

`gemm()` computes C=αop(A)op(B)+βC, with β=0 and C being columns set here, so the initial value of columns does not matter and it is used as an output.

This is also verified by gradcheck passing when using `at::empty()`.

jbschlosser commented on aten/src/ATen/native/NaiveConvolutionTranspose3d.cpp:
```diff
@@ -311,7 +295,7 @@ void slow_conv_transpose3d_out_cpu_template(
           const int64_t m =
               weight.size(1) * weight.size(2) * weight.size(3) * weight.size(4);
-          const int64_t n = columns.size(1);
+          const int64_t n = input_depth * input_height * input_width;
```
I don't strictly need to change it - this was just for consistency with the backward. Since columns is unconditionally being allocated above, I'll change it back

jbschlosser commented on aten/src/ATen/native/cuda/NaiveConvolutionTranspose3d.cu:
```diff
@@ -314,7 +300,7 @@ void slow_conv_transpose3d_out_cuda_template(
           int64_t m =
               weight.size(1) * weight.size(2) * weight.size(3) * weight.size(4);
-          int64_t n = columns.size(1);
+          int64_t n = input_depth * input_height * input_width;
```
Nope not strictly needed. It was made for the backward pass because sometimes columns does not need to be allocated. I'll revert the change here

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).

jbschlosser committed:
Update on "Remove finput and fgrad_input from slow3d transpose signatures"


Differential Revision: [D32655872](https://our.internmc.facebook.com/intern/diff/D32655872)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32655872).
